x/epoching/keeper/msg_server.go

```diff
@@ -16,13 +18,135 @@ type msgServer struct {
 	Keeper
 }
 
+var _ types.MsgServer = msgServer{}
+
 // NewMsgServerImpl returns an implementation of the MsgServer interface
 // for the provided Keeper.
 func NewMsgServerImpl(keeper Keeper) types.MsgServer {
 	return &msgServer{Keeper: keeper}
 }
 
-var _ types.MsgServer = msgServer{}
+// WrappedEditValidator handles the MsgWrappedEditValidator request
+func (ms msgServer) WrappedEditValidator(goCtx context.Context, msgWrapped *types.MsgWrappedEditValidator) (*types.MsgWrappedEditValidatorResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	if msgWrapped.Msg == nil {
+		return nil, types.ErrNoWrappedMsg
+	}
+	msg := msgWrapped.Msg
+
+	// verification rules ported from staking module
+	valAddr, valErr := sdk.ValAddressFromBech32(msg.ValidatorAddress)
+	if valErr != nil {
+		return nil, valErr
+	}
+
+	if msg.Description == (stktypes.Description{}) {
+		return nil, errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "empty description")
+	}
+
+	if msg.MinSelfDelegation != nil && !msg.MinSelfDelegation.IsPositive() {
+		return nil, errorsmod.Wrap(
+			sdkerrors.ErrInvalidRequest,
+			"minimum self delegation must be a positive integer",
+		)
+	}
+
+	if msg.CommissionRate != nil {
+		if msg.CommissionRate.GT(math.LegacyOneDec()) || msg.CommissionRate.IsNegative() {
+			return nil, errorsmod.Wrap(sdkerrors.ErrInvalidRequest, "commission rate must be between 0 and 1 (inclusive)")
+		}
+
+		stkParams, err := ms.stk.GetParams(ctx)
+		if err != nil {
+			return nil, errorsmod.Wrap(sdkerrors.ErrLogic, err.Error())
+		}
+
+		minCommissionRate := stkParams.MinCommissionRate
+
+		if msg.CommissionRate.LT(minCommissionRate) {
+			return nil, errorsmod.Wrapf(sdkerrors.ErrInvalidRequest, "commission rate cannot be less than the min commission rate %s", minCommissionRate.String())
+		}
+	}
+
+	// validator must already be registered
+	_, err := ms.stk.GetValidator(ctx, valAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	blockHeight := uint64(ctx.HeaderInfo().Height)
+	if blockHeight == 0 {
+		return nil, types.ErrZeroEpochMsg
+	}
+	blockTime := ctx.HeaderInfo().Time
+
+	txid := tmhash.Sum(ctx.TxBytes())
+	queuedMsg, err := types.NewQueuedMessage(blockHeight, blockTime, txid, msg)
+	if err != nil {
+		return nil, err
+	}
+
+	ms.EnqueueMsg(ctx, queuedMsg)
```

Comment:
Why do we need to defer the `EditValidator` operation to the end of the epoch, rather than executing it directly?

Reply:
That is a good question, I also was thinking about it when writing it, but I thought since it could change the min self delegation amount and the commission was better to defer to the end of epoch

What do you think?